### PR TITLE
Combine translations

### PR DESCRIPTION
**Pull request by Amr-Nash**
Combine translations files from all languages to one file for each

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+	make clean_translations_temp_directory
+	atlas pull $(ATLAS_OPTIONS) translations/openedx-app-ios/I18N/I18N/:I18N/I18N/
+	python i18n_scripts/split_translation_files.py
+	make clean_translations_temp_directory
+
+
+combine_translations:
+	make clean_translations_temp_directory
+	pip3 install localizable==0.1.2
+	python3 i18n_scripts/combine_translations_files.py
```
Comment (OmarIthawi) [inline Makefile]: Please add the following lines:

```suggestion
create_virtual_env:
	rm -rf .venv
	python3 -m venv .venv
	. .venv/bin/activate && pip install localizable==0.1.2
```

combine_translations: create_virtual_env
	make clean_translations_temp_directory
	. .venv/bin/activate && python3 i18n_scripts/combine_translations_files.py
```

File: i18n_scripts/combine_translations_files.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
+import os, localizable
+from collections import OrderedDict
+
+
+def translation_file_of(module):
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]: Add docustring here

```suggestion
def get_translation_file_path(module):
    """
    Return translation path from module name.
    """
```

File: i18n_scripts/combine_translations_files.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    """
+    Retrieve the names of modules that have translation files for a specified language.
+
+    Parameters:
+        modules_dir (str): The path to the directory containing all the modules.
+
+    Returns:
+        list of str: A list of module names that have translation files for the specified language.
+    """
+    dirs = [directory for directory in os.listdir(modules_dir) if
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]: reformat in multiple lines

File: i18n_scripts/combine_translations_files.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            os.path.isdir(os.path.join(modules_dir, directory))]
+
+    modules_list = []
+    for module in dirs:
+        translation_file = translation_file_of(module)
+        if os.path.exists(translation_file) and os.path.isfile(translation_file):
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]: this should be enough:

```suggestion
        if os.path.isfile(translation_file):
```

File: i18n_scripts/combine_translations_files.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            ordered_dict_of_translations[f"{module}.{line['key']}"] = line
+
+    return ordered_dict_of_translations
+
+
+def _escape(s):
+    """
+    Reverse the replacements performed by _unescape() in the localizable library
+    """
+    s = s.replace('\n', r'\n').replace('\r', r'\r').replace('"', r'\"')
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]: Please use `re.sub` to include `\\` as well:

```suggestion
    s = re.sub(r'\n|\r|"|\\', lambda match: f'\\{match.group(0)}')
```

I've written this on GitHub, but I think it'll work. Please test it.

File: i18n_scripts/combine_translations_files.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+       This function takes an ordered dictionary containing translation data and writes it to a Localizable.strings
+       file located in the 'I18N/en.lproj' directory within the specified modules directory. It creates the directory
+       if it doesn't exist.
+
+       Parameters:
+           content_ordered_dict (OrderedDict): An ordered dictionary containing translation data. The keys
+           are the translation keys, and the values are dictionaries with 'value' and 'comment' keys representing the
+           translation value and optional comments, respectively.
+           modules_dir (str): The path to the modules directory
+           where the I18N directory will be written.
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]: Please de-indent by 3-spaces to match the start of `"""`

File: i18n_scripts/combine_translations_files.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+def _escape(s):
+    """
+    Reverse the replacements performed by _unescape() in the localizable library
+    """
+    s = s.replace('\n', r'\n').replace('\r', r'\r').replace('"', r'\"')
+    return s
+
+
+def write_dict_to_file(content_ordered_dict, modules_dir):
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]:
```suggestion
def write_translations_file(translations_dict, modules_dir):
```

File: i18n_scripts/combine_translations_files.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+            comment = value.get('comment', '')  # Retrieve the comment, if present
+            if comment:
+                f.write(f"/* {comment} */\n")
+            f.write(f'"{key}" = "{_escape(value["value"])}";\n')
+
+
+if __name__ == "__main__":
+    modules_dir = os.path.dirname(os.path.dirname(__file__))
+    en_translation_ordered_dict = get_translations(modules_dir)
+    write_dict_to_file(en_translation_ordered_dict, modules_dir)
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]: Please refactor to:

```suggestion
def combine_translation_files():
    """
    ADD DOCS HERE.
    """
    modules_dir = os.path.dirname(os.path.dirname(__file__))
    en_translation_ordered_dict = get_translations(modules_dir)
    write_dict_to_file(en_translation_ordered_dict, modules_dir)

if __name__ == "__main__":
    combine_translation_files()
```

File: i18n_scripts/combine_translations_files.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+
+def make_test_dirs_and_files():
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]: Please remove this function for now.

We need to refactor testing or submit without testing.

**Review by OmarIthawi — COMMENTED** [review]
Thanks a lot @Amr-Nash! 

I've added multiple comments.

Please address them, test and let me know when it's ready.

File: i18n_scripts/combine_translations_files.py (hunk 2 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: i18n_scripts/combine_translations_files.py (hunk 4 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: i18n_scripts/combine_translations_files.py (hunk 5 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: fixed the escape function using re.sub

File: i18n_scripts/combine_translations_files.py (hunk 6 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: i18n_scripts/combine_translations_files.py (hunk 7 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: i18n_scripts/combine_translations_files.py (hunk 9 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: i18n_scripts/combine_translations_files.py (hunk 8 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: Makefile (hunk 1 above)
Comment (Amr-Nash) [reply Makefile]: done for now

File: i18n_scripts/combine_translations_files.py [hunk 10]
```diff
@@ -71,11 +71,10 @@ def _escape(s):
     replacement_function = lambda match: {
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]: make it 


```suggestion
    def escape_char(match):
```

File: i18n_scripts/combine_translations_files.py [hunk 11]
```diff
@@ -71,11 +71,10 @@ def _escape(s):
...
         '\n': r'\n',
         '\r': r'\r',
-        '\\': r'\\',
         '"': r'\"'
     }[match.group(0)]
 
-    s = re.sub(r'([\n\r\\"])', replacement_function, s)
+    s = re.sub(r'([\n\r"])', replacement_function, s)
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]: does it work on multiple characters in the same string?

`yes """" it's me`

File: i18n_scripts/combine_translations_files.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    Retrieve the names of modules that have translation files for a specified language.
+
+    Parameters:
+        modules_dir (str): The path to the directory containing all the modules.
+
+    Returns:
+        list of str: A list of module names that have translation files for the specified language.
+    """
+    dirs = [directory for directory in os.listdir(modules_dir) if
+            os.path.isdir(os.path.join(modules_dir, directory))]
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]:
```suggestion
    dirs = [
        directory for directory in os.listdir(modules_dir)
        if os.path.isdir(os.path.join(modules_dir, directory))
    ]
```

File: i18n_scripts/combine_translations_files.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+    for module in dirs:
+        translation_file = get_translation_file_path(modules_dir, module)
+        if os.path.isfile(translation_file):
+            modules_list.append(module)
+    return modules_list
+
+
+def get_translations(modules_dir):
+    """
+    Retrieve the list of translations from all specified modules.
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]:
```suggestion
    Retrieve the translations from all specified modules as OrderedDict.
```

File: i18n_scripts/combine_translations_files.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        translation line. The key of the OrderedDict is the value of the key in the contained dict in addition to the
+        name of the module containing the translation
+    """
+    ordered_dict_of_translations = OrderedDict()
+    modules = get_modules_to_translate(modules_dir)
+    for module in modules:
+        translation_file = get_translation_file_path(modules_dir, module)
+        module_translations = localizable.parse_strings(filename=translation_file)
+
+        for line in module_translations:
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]:
```suggestion
        for entry in module_translations:
```

File: i18n_scripts/combine_translations_files.py [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+        '\n': r'\n',
+        '\r': r'\r',
+        '"': r'\"'
+    }[match.group(0)]
+
+    s = re.sub(r'([\n\r"])', replacement_function, s)
+    return s
+
+
+def write_translations_file(content_ordered_dict, modules_dir):
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]:
```suggestion
def write_translation_file(content_ordered_dict, modules_dir):
```

File: i18n_scripts/combine_translations_files.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+    if it doesn't exist.
+    
+    Parameters:
+       content_ordered_dict (OrderedDict): An ordered dictionary containing translation data. The keys
+       are the translation keys, and the values are dictionaries with 'value' and 'comment' keys representing the
+       translation value and optional comments, respectively.
+       modules_dir (str): The path to the modules directory
+       where the I18N directory will be written.
+    """
+    combined_translation_file = os.path.join(modules_dir, 'I18N', 'en.lproj')
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]:
```suggestion
    combined_translation_dir = os.path.join(modules_dir, 'I18N', 'en.lproj')
```

File: i18n_scripts/combine_translations_files.py [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+    os.makedirs(combined_translation_file, exist_ok=True)
+    with open(os.path.join(combined_translation_file, 'Localizable.strings'), 'w') as f:
+        for key, value in content_ordered_dict.items():
+            comment = value.get('comment', '')  # Retrieve the comment, if present
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]:
```suggestion
            comment = value.get('comment')  # Retrieve the comment, if present
```

File: i18n_scripts/combine_translations_files.py [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+            if comment:
+                f.write(f"/* {comment} */\n")
+            f.write(f'"{key}" = "{_escape(value["value"])}";\n')
+
+
+def combine_translation_files():
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]:
```suggestion
def combine_translation_files(modules_dir=None):
```

File: i18n_scripts/combine_translations_files.py [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+    """
+    Combine translation files from different modules into a single file.
+    """
+    modules_dir = os.path.dirname(os.path.dirname(__file__))
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]:
```suggestion
    if not modules_dir:
       modules_dir = os.path.dirname(os.path.dirname(__file__))
```

File: i18n_scripts/combine_translations_files.py [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+    en_translation_ordered_dict = get_translations(modules_dir)
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]:
```suggestion
    combined_translation_dict = get_translations(modules_dir)
```

**Review by OmarIthawi — COMMENTED** [review]
some nit-picks, otherwise it looks good

File: i18n_scripts/combine_translations_files.py (hunk 10 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: Done

File: i18n_scripts/combine_translations_files.py (hunk 15 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: i18n_scripts/combine_translations_files.py (hunk 14 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: i18n_scripts/combine_translations_files.py (hunk 13 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: i18n_scripts/combine_translations_files.py (hunk 12 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: i18n_scripts/combine_translations_files.py (hunk 11 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: Yes, it does.

File: i18n_scripts/combine_translations_files.py (hunk 16 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: i18n_scripts/combine_translations_files.py (hunk 17 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: i18n_scripts/combine_translations_files.py (hunk 18 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: i18n_scripts/combine_translations_files.py (hunk 19 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: i18n_scripts/combine_translations_files.py (hunk 20 above)
Comment (Amr-Nash) [reply i18n_scripts/combine_translations_files.py]: done

File: i18n_scripts/combine_translations_files.py [hunk 21]
```diff
@@ -3,6 +3,11 @@
 import localizable
 from collections import OrderedDict
 
+"""
+Translations are pulled from https://github.com/openedx/openedx-translations to the I18N directory.
+This script splits the translations from the I18N directory to all modules in the repository.
+"""
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]:
```suggestion
"""
Translations are pulled from https://github.com/openedx/openedx-translations to the I18N directory.
This script splits the translations from the I18N directory to all modules in the repository.
"""

import localizable
from collections import OrderedDict

```

File: i18n_scripts/combine_translations_files.py [hunk 22]
```diff
@@ -0,0 +1,122 @@
+import os
+import re
+import localizable
+from collections import OrderedDict
+
+"""
+This script combines the English translations from all modules in the repository to the I18N directory.
+Then the translations are pushed to the repo https://github.com/openedx/openedx-translations from the I18N directory.
+"""
```
Comment (OmarIthawi) [inline i18n_scripts/combine_translations_files.py]:
```suggestion
"""
This script combines the English translations from all modules in the repository to the I18N directory.
Then the translations are pushed to the repo https://github.com/openedx/openedx-translations from the I18N directory.
"""

import os
import re
import localizable
from collections import OrderedDict

```

**Review by OmarIthawi — APPROVED** [review]
@Amr-Nash Thanks for this. Please put the docs above all lines.

**Pull request merged**